core/domain/story_services.py

```diff
@@ -101,16 +101,28 @@ def apply_change_list(story_id, change_list):
             story.
 
     Returns:
-        Story. The resulting story domain object.
+        Story, list(str), list(str). The resulting story domain object, the
+            exploration IDs removed from story and the exploration IDs added to
+            the story.
     """
     story = story_fetchers.get_story_by_id(story_id)
+    exp_ids_removed_from_story = []
+    exp_ids_added_to_story = []
     try:
         for change in change_list:
             if not isinstance(change, story_domain.StoryChange):
                 raise Exception('Expected change to be of type StoryChange')
             if change.cmd == story_domain.CMD_ADD_STORY_NODE:
                 story.add_node(change.node_id, change.title)
             elif change.cmd == story_domain.CMD_DELETE_STORY_NODE:
+                for node in story.story_contents.nodes:
+                    if node.id == change.node_id:
+                        if node.exploration_id in exp_ids_added_to_story:
```

Comment:
In the frontend itself, the 'Save' button is disabled if duplicate explorations are detected. So, in normal use, the only way same explorations would be present in multiple chapters is if it is removed from the first chapter before adding it to the second chapter, does that resolve the question?

Reply:
I thought the validation check in the frontend is only applied for the entire set of changes? So, it is possible to make the above sequence of changes and then click publish, which means that you're still going to run into that issue here when you iterate through the changes one by one, right?

Reply:
No, the 'Save' button beside the exploration ID field itself would raise an error if the same exploration ID is already present. So, a changelist that has duplicate explorations in the story at some point is not generated.

Reply:
Ah, ok, thanks. But I think we still should do it "properly" in the backend, since the frontend may change in the future... in fact, I thought we had a discussion in the LaCE team or in another PR where there was a suggestion to move this check to the navbar?

Reply:
Yeah, actually on checking, when the `update_node_exploration_id` function is called in the backend also, the update doesn't happen if the exploration already exists in the story. So, the validation error would be raised there. Wouldn't that be enough?

Reply:
As long as it remains impossible for the frontend to send a "bad" changelog (incorporating duplicate exp IDs) to the backend, I think that sounds OK. But see my other comment below -- another idea just occurred to me which I think might be much better, unless I'm missing something.